### PR TITLE
Draw overlay correctly in HDR

### PR DESCRIPTION
**Pull request by EndlesslyFlowering**
Content of this PR:
- add two new shaders ([`imgui_ps_4_0_hdr.hlsl`](https://github.com/crosire/reshade/pull/261/commits/922909e85ab9dc7663866267b2e3b4642bc717c6#diff-a1afdcdb56dcf865da8310ecd6be1f8fbfd3c03bfbd672e669c329b01f09854e) and [`imgui_ps_hdr.glsl`](https://github.com/crosire/reshade/pull/261/commits/922909e85ab9dc7663866267b2e3b4642bc717c6#diff-4b551b433e8da07812f32169f09b18ee0be4305bd6c4dfcb6e2160ea95669820)) for drawing ReShade's imgui overlay correctly in HDR which are only used for possible HDR swapchains
- add [two new variables](https://github.com/crosire/reshade/pull/261/commits/922909e85ab9dc7663866267b2e3b4642bc717c6#diff-5c5309799d91aef1e44123aa797c5afe37dfc32c6c19ff16b7d34531f50b0f6dR511-R516) in a struct: `_hdr_overlay.brightness` to control the brightness of the overlay in HDR and `_hdr_overlay.overwrite_color_space_to` if the auto detection fails (this can happen for early HDR games and all UE4 games as they rely on AMD AGS and/or NvAPI to change the display's color space as `IDXGISwapChain3::SetColorSpace1` wasn't a thing yet)
- ~~pack `_back_buffer_format` and `_back_buffer_color_space` into a [struct](https://github.com/crosire/reshade/pull/261/commits/922909e85ab9dc7663866267b2e3b4642bc717c6#diff-5c5309799d91aef1e44123aa797c5afe37dfc32c6c19ff16b7d34531f50b0f6dR188-R192) so `push_constants` doesn't need to be called too often (ideally both the `_hdr_overlay` and the `_back_buffer` struct would be consolidated into one single struct)~~
- add ability to call `push_constants` [multiple times](https://github.com/crosire/reshade/pull/261/commits/2edea25c8fcd37481f5ed0f906347c71dc23820f) for d3d11 by always keeping a copy of the push constants `pData` around (needed for the HDR overlay settings to be added to the push constants after the `ortho_projection` has been added first)
- ~~add `*.spv` files to `.gitignore` and delete all `*.spv` files~~

see latest comment why things are crossed out.

### CODE REVIEW

**Comment by EndlesslyFlowering** [comment]
gonna split this into 3 commits:
- hdr overlay stuff
- add ability to call push_constants in d3d11
- add *.spv to .gitignore and delete all *.spv files

**Comment by EndlesslyFlowering** [comment]
done
also updated the opening post to reflect the changes

**Comment by EndlesslyFlowering** [comment]
- dropped the *.spv files ignore commit and created a separate PR for that [here](https://github.com/crosire/reshade/pull/263)
- undid the packing of `_back_buffer_format` and `_back_buffer_color_space` into a struct in exchange for packing all needed variables for hdr overlay into struct before calling `push_constants` so it's only needed to be called once

File: source/d3d11/d3d11_impl_command_list.cpp [hunk 1]
```diff
@@ -396,18 +396,38 @@ void reshade::d3d11::device_context_impl::push_constants(api::shader_stage stage
...
+		_old_push_constants_data = nullptr;
+	}
+
+	if (new_size > _push_constants_size)
 	{
 		// Enlarge push constant buffer to fit new requirement
 		D3D11_BUFFER_DESC desc = {};
-		desc.ByteWidth = count * sizeof(uint32_t);
+		// ByteWidth has to be a multiple of 16
+		desc.ByteWidth = static_cast<uint32_t>(round(ceilf(new_size * sizeof(uint32_t) / 16.f) * 16.f));
```
Comment (crosire) [inline source/d3d11/d3d11_impl_command_list.cpp:419]: Quicker way to solve this without having to involve floating point math:
```c++
desc.ByteWidth = ((new_size * sizeof(uint32_t)) + 15) & ~15;
```

File: source/d3d11/d3d11_impl_command_list.cpp [hunk 2]
```diff
@@ -396,18 +396,38 @@ void reshade::d3d11::device_context_impl::push_constants(api::shader_stage stage
...
 		desc.Usage = D3D11_USAGE_DYNAMIC;
 		desc.BindFlags = D3D11_BIND_CONSTANT_BUFFER;
 		desc.CPUAccessFlags = D3D11_CPU_ACCESS_WRITE;
 
+		// Release current push constants so we can create a new buffer with a different size
+		if (_push_constants.get() != nullptr)
+		{
+			_push_constants->Release();
+			_push_constants.release();
+		}
```
Comment (crosire) [inline source/d3d11/d3d11_impl_command_list.cpp:429]: These 5 lines can be replaced with a simple one liner:
```c++
_push_constants.reset();
```

File: source/runtime_gui.cpp [hunk 3]
```diff
@@ -285,6 +285,8 @@ void reshade::runtime::load_config_gui(const ini_file &config)
 	config.get("STYLE", "FPSScale", _fps_scale);
 	config.get("STYLE", "FrameRounding", imgui_style.FrameRounding);
 	config.get("STYLE", "GrabRounding", imgui_style.GrabRounding);
+	config.get("STYLE", "HdrOverlayBrightness", _hdr_overlay.brightness);
+	config.get("STYLE", "HdrOverlayOverwriteColorSpaceTo", _hdr_overlay.overwrite_color_space_to);
```
Comment (crosire) [inline source/runtime_gui.cpp:289]: I'm debating whether these are best kept in the `STYLE` or `OVERLAY` section ... Mmh.

File: source/runtime_gui.cpp [hunk 4]
```diff
@@ -2114,6 +2118,23 @@ void reshade::runtime::draw_gui_settings()
...
 		}
 
+		// only for d3d11, d3d12, Vulkan API and possible hdr swapchains
+		if ((_renderer_id & 0xB000 || _renderer_id & 0xC000 || _renderer_id & 0x20000)
+		 && (_back_buffer_format == reshade::api::format::r10g10b10a2_unorm || _back_buffer_format == reshade::api::format::b10g10r10a2_unorm ||_back_buffer_format == reshade::api::format::r16g16b16a16_float))
+		{
+			if (float &hdr_overlay_brightness = _hdr_overlay.brightness; ImGui::SliderFloat("HDR overlay brightness", &hdr_overlay_brightness, 20.f, 400.f, "%.0f nits"))
+			{
+				// Prevent user from setting the brightness too low
+				hdr_overlay_brightness = std::max(hdr_overlay_brightness, 20.f);
```
Comment (crosire) [inline source/runtime_gui.cpp:2128]: ImGui has this nice new flag `ImGuiSliderFlags_AlwaysClamp`, which ensures users can't input values below the specified range. That can probably replace this explicit clamp here.

File: source/runtime_gui.cpp [hunk 5]
```diff
@@ -4000,7 +4021,18 @@ bool reshade::runtime::init_imgui_resources()
...
 
-		layout_params[num_layout_params++] = api::constant_range { 0, 0, 0, 16, api::shader_stage::vertex }; // b0
+		uint32_t count = 16;
+		reshade::api::shader_stage shader_stage = api::shader_stage::vertex;
+
+		// only for d3d11, d3d12, Vulkan API and possible hdr swapchains
+		if ((_renderer_id & 0xB000 || _renderer_id & 0xC000 || _renderer_id & 0x20000)
+		 && (_back_buffer_format == reshade::api::format::r10g10b10a2_unorm || _back_buffer_format == reshade::api::format::b10g10r10a2_unorm ||_back_buffer_format == reshade::api::format::r16g16b16a16_float))
+		{
+			count = 20;
```
Comment (crosire) [inline source/runtime_gui.cpp:4031]: For sake of readability, maybe `count += 4;` instead?

File: res/shaders/imgui_ps_4_0_hdr.hlsl [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+//scRGB (linear + BT.709/sRGB primaries)
+float4 to_scrgb(float4 col)
+{
+	return float4(pow(col.rgb, 2.2f) * hdr_overlay_brightness / 80.f, col.a);
+}
+
+void main(float4 vpos : SV_POSITION, float4 vcol : COLOR0, float2 uv : TEXCOORD0, out float4 col : SV_TARGET)
+{
+	col = t0.Sample(s0, uv);
```
Comment (crosire) [inline res/shaders/imgui_ps_4_0_hdr.hlsl:55]: I'm curious how textures interact with this. They will be in sRGB color space coming out of here and only the color space of the overlay colors mixed into that below are changed. Does that hold up? E.g. what happens when looking at the Statistics tab in HDR.

**Review by crosire — COMMENTED** [review]
This is great to have! I don't have the hardware to actually try this unfortunately, so can only really look at the code.

**Comment by crosire** [comment]
Seems also relevant to consider https://github.com/SpecialKO/SpecialK/commit/1f590f68c6e2c6f25b3704926a250d1735d96afc.

**Comment by EndlesslyFlowering** [comment]
> For reference [SpecialKO/SpecialK@1f590f6](https://github.com/SpecialKO/SpecialK/commit/1f590f68c6e2c6f25b3704926a250d1735d96afc) may also be relevant.

thanks! SpecialK will not need that anymore in the future :D

just applied all the issues brought up and recheck everything.

also added HLG for completeness sake to the shaders.

**Review by crosire — APPROVED** [review]
Looking good. I'll port over the push constant changes to D3D10 in a follow up commit (and might change the implementation slightly).

**Comment by EndlesslyFlowering** [comment]
uh about the license headers in the HDR specific pixel shaders. I wasn't sure how to add them. so they aren't in them yet.

**Pull request merged**